### ISSUE TITLE
[Bug]: SSL_read in streaming/receiver.c not checking for disconnection, 100% CPU

### ISSUE THREAD

**Issue by ZaneA** [netdata/netdata#13092]
### Bug description

When agents disconnect from a Netdata server after being connected via SSL, we found the server burning 100% of a CPU core. This happened after upgrading the Netdata server (past 1.32.x I believe), reverting back to an older version would fix things, until attempting to upgrade again. The agent version would be relatively recent (e.g. 1.34.x).

`strace` pointed to a `read` system call in a tight loop, always returning 0. I narrowed it down to this line in the new streaming code, introduced a couple of releases back which calls `SSL_read` in a loop while the return value is 0 (https://github.com/netdata/netdata/blob/master/streaming/receiver.c#L222).

Looking at the OpenSSL docs (https://www.openssl.org/docs/man1.1.1/man3/SSL_read.html), `SSL_get_error()` should be called for a return value <= 0 to determine the reason. In our case the connection was closed by this point, as opposed to just waiting for more data, so the loop never returns.

I could reliably reproduce this with SSL enabled, once disabling SSL in the agent destination and bypassing this code I was unable to reproduce the issue again.

### Expected behavior

Netdata should continue to function normally when agents are disconnected.

### Steps to reproduce

1. Enable SSL on the Netdata server and agent
2. Run the agent (should be on a recent release, >1.32?)
3. Stop the agent
4. The server should start spinning

### Installation method

docker

### System info

```shell
Linux xxxx 5.4.0-117-generic #132-Ubuntu SMP Thu Jun 2 00:39:06 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
Ubuntu 20.04
```

### Netdata build info

```shell
We have since upgraded to 1.35.0 - however I believe this bug still exists in 1.35.0 (untested).

Version: netdata v1.35.0
Configure options:  '--build=x86_64-linux-gnu' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--disable-silent-rules' '--libdir=${prefix}/lib/x86_64-linux-gnu' '--libexecdir=${prefix}/lib/x86_64-linux-gnu' '--disable-maintainer-mode' '--prefix=/usr' '--sysconfdir=/etc' '--localstatedir=/var' '--libdir=/usr/lib' '--libexecdir=/usr/libexec' '--with-user=netdata' '--with-math' '--with-zlib' '--with-webdir=/var/lib/netdata/www' '--disable-dependency-tracking' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -fdebug-prefix-map=/usr/src/netdata=. -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-Bsymbolic-functions -Wl,-z,relro' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -fdebug-prefix-map=/usr/src/netdata=. -fstack-protector-strong -Wformat -Werror=format-security'
Install type: binpkg-deb
    Binary architecture: x86_64
    Packaging distro:  
Features:
    dbengine:                   YES
    Native HTTPS:               YES
    Netdata Cloud:              YES 
    ACLK Next Generation:       YES
    ACLK-NG New Cloud Protocol: YES
    ACLK Legacy:                NO
    TLS Host Verification:      YES
    Machine Learning:           YES
    Stream Compression:         YES
Libraries:
    protobuf:                YES (system)
    jemalloc:                NO
    JSON-C:                  YES
    libcap:                  NO
    libcrypto:               YES
    libm:                    YES
    tcalloc:                 NO
    zlib:                    YES
Plugins:
    apps:                    YES
    cgroup Network Tracking: YES
    CUPS:                    YES
    EBPF:                    YES
    IPMI:                    YES
    NFACCT:                  YES
    perf:                    YES
    slabinfo:                YES
    Xen:                     NO
    Xen VBD Error Tracking:  NO
Exporters:
    AWS Kinesis:             NO
    GCP PubSub:              NO
    MongoDB:                 NO
    Prometheus Remote Write: YES
```

**Comment by MrZammler** [comment]
Hi @ZaneA ! Thank you for the report, we will try to reproduce and fix!

**Comment by thiagoftsm** [comment]
Hello @ZaneA ,

Thank you very much for your description. I was able to replicate the issue that you described right now. I observed that `netdata` is entering in an infinite loop and `SSL_get_error` is returning `SSL_ERROR_SYSCALL`. Considering that I am a long time without to work with this specific code, I am stopping right now, because it is too late in my timezone, and when I start working in few hours, I will do a complete review  in the `receiver` code and I will also test using different `OpenSSL` versions to be sure that the solution that I tested now is enough to fix everything.

Before I go, congratulation by your tests and precise description of the issue!

Best Regards!
